### PR TITLE
allow setting multiple matchers to "promtool tsdb dump"

### PR DESCRIPTION
**Pull request by machine424**
add unit tests for "promtool tsdb dump".

refactor some matchers scraping utils.

fixes https://github.com/prometheus/prometheus/issues/13295

### CODE REVIEW

**Review by bboreham — APPROVED** [review]
Code change looks fine; could you put something in the description about why you want this?

**Comment by bboreham** [comment]
Sorry I realise it’s in the issue: “ To get more granularity without having to run the command multiple times.”

However this leaves me unclear. Do the matchers AND?

**Comment by machine424** [comment]
> However this leaves me unclear. Do the matchers AND?

Conditions are ANDed inside the same matcher but matchers are ORed, like the `match[]` for the query API (See the added test `well merged`).

File: cmd/promtool/tsdb_test.go [hunk 1]
```diff
@@ -41,3 +50,101 @@ func TestGenerateBucket(t *testing.T) {
...
+// getDumpedSamples dumps samples and returns them.
+func getDumpedSamples(t *testing.T, path string, mint, maxt int64, match []string) string {
+	t.Helper()
+
+	oldStdout := os.Stdout
+	r, w, _ := os.Pipe()
+	os.Stdout = w
+
+	err := dumpSamples(
+		context.TODO(),
```
[the remark this answers is not in the document]
Comment (machine424) [reply cmd/promtool/tsdb_test.go:63]: Ah, a `Background` would be sufficient for such a test, will change, thanks.

**Pull request merged**